pyth-sdk/src/price.rs

```diff
@@ -89,6 +91,262 @@ impl Price {
+    /// y1: Price, the y coordinate of the first point, represented as a Price struct
+    /// x2: i64, the x coordinate of the second point, must be greater than x1
+    /// y2: Price, the y coordinate of the second point, represented as a Price struct
+    /// x_query: i64, the query x coordinate, at which we wish to impute a y value
+    /// pre_add_expo: i32, the exponent to scale to, before final addition; essentially the final precision you want
+    /// 
+    /// Logic
+    /// imputed y value = y2 * ((xq-x1)/(x2-x1)) + y1 * ((x2-x3)/(x2-x1))
+    /// 1. compute A = xq-x1
+    /// 2. compute B = x2-xq
+    /// 3. compute C = x2-x1
+    /// 4. compute D = A/C
+    /// 5. compute E = B/C
+    /// 6. compute F = y2 * D
+    /// 7. compute G = y1 * E
+    /// 8. compute H = F + G
+    /// 
+    /// Bounds due to precision loss
+    /// x = 10^(PD_EXPO+2)
+    /// fraction (due to normalization & division) incurs max loss of x
+    /// Thus, max loss here: Err(D), Err(E) <= x
+    /// If y1, y2 already normalized, no additional error. O/w, Err(y1), Err(y2) with normalization <= x
+    /// Err(F), Err(G) <= (1+x)^2 - 1 (in fractional terms) ~= 2x
+    /// Err(H) <= 2*2x = 4x, when PD_EXPO = -9 ==> Err(H) <= 4*10^-7
+    /// 
+    /// Scaling this back has error bounded by the expo (10^pre_add_expo).
+    /// This is because reverting a potentially finer expo to a coarser grid has the potential to be off by
+    /// the order of the atomic unit of the coarser grid.
+    /// This scaling error combines with the previous error additively: Err <= 2/PD_SCALE + 2*10^pre_add_expo
+    /// But if pre_add_expo is reasonably small (<= -9), then other term will dominate
+    pub fn affine_combination(x1: i64, y1: Price, x2: i64, y2: Price, x_query: i64, pre_add_expo: i32) -> Option<Price> {
+        if x2 <= x1 {
+            return None;
+        }
+        
+        // get the deltas for the x coordinates
+        // 1. compute A = xq-x1
+        let delta_q1 = x_query.checked_sub(x1)?;
+        // 2. compute B = x2-xq
+        let delta_2q = x2.checked_sub(x_query)?;
+        // 3. compute C = x2-x1
+        let delta_21 = x2.checked_sub(x1)?;
+
+        // get the relevant fractions of the deltas, with scaling
+        // 4. compute D = A/C, Err(D) <= x = 1/PD_SCALE
+        let frac_q1 = Price::fraction(delta_q1, delta_21)?;
+        // 5. compute E = B/C, Err(E) <= x
+        let frac_2q = Price::fraction(delta_2q, delta_21)?;
+
+        // calculate products for left and right
+        // 6. compute F = y2 * D, Err(F) <= (1+x)^2 - 1
+        let mut left = y2.mul(&frac_q1)?;
+        // 7. compute G = y1 * E, Err(G) <= (1+x)^2 - 1
+        let mut right = y1.mul(&frac_2q)?; 
+
+        // Err(scaling) += 2*10^pre_add_expo
+        left = left.scale_to_exponent(pre_add_expo)?;
+        right = right.scale_to_exponent(pre_add_expo)?;
+        
+        if left.expo != right.expo {
+            return None;
+        }
```

Comment:
delete this if statement (it can't happen since you just scaled them to the same expo above)

Reply:
you can still delete this if statement